### PR TITLE
DCMAW-10014: JWKS endpoint for public encryption key

### PR DESCRIPTION
**Pull request by jmooney-dd**
​DCMAW-10014

Publish the ID check public encryption key to the `/.well-known/jwks.json` endpoint

### Please note

This change was originally merged in #99 however [failed on the GHA](https://github.com/govuk-one-login/mobile-id-check-async/actions/runs/10770081064/job/29862876609) due to template linting. The PR was reverted and this is now the revert, of the revert, with template fixes

### What changed
- Added `/.well-known/jwks.json` endpoint to the public OpenAPI specification
- Added infrastructure (JWKS Lambda, JWKS S3, JWKS Lambda Custom Resource, log groups, policies, IAM roles)
- JWKS Lambda function code
  - Handler
  - JWKS Builder class
  - JWKS Upload class
  - Custom Resource Event Sender class

Stack deployed to the `dev` environment

Screenshot 2024-08-30 at 12 43 54

Linting template now passes:

Screenshot 2024-09-09 at 15 52 30

### Why did it change
- STS needs the ID check public encryption key to encrypt the service token they issue

## Checklists

- [x] There is a ticket raised for this PR that is present in the branch name
- [x] No PII data logged. [See guidance here](https://govukverify.atlassian.net/wiki/spaces/DCMAW/pages/3502407722/PII+Logging+Considerations)
- [ ] Demo to a BA, TA, and the team.
- [x] Update [README](./blob/main/README.md) with any new instructions or tasks

### CODE REVIEW

File: backend-api/src/functions/tests/api/jwks.test.ts
Comment (sandymay-dd) [inline backend-api/src/functions/tests/api/jwks.test.ts:1]: please may we move this into a top level folder (backend-api/api-tests)?

File: backend-api/package.json [hunk 1]
```diff
@@ -8,10 +8,11 @@
     "format:check": "./node_modules/.bin/prettier --check '**/*.ts'",
     "lint": "./node_modules/.bin/eslint ./src/functions",
     "pact": "./node_modules/.bin/jest --testPathPattern=tests/pact",
-    "test": "./node_modules/.bin/jest",
+    "test": "./node_modules/.bin/jest --testPathIgnorePatterns=tests/api",
+    "test:api": "./node_modules/.bin/jest --testPathPattern=tests/api",
     "test:infra": "./node_modules/.bin/jest --testPathPattern=infra-tests/",
     "test:pact": "npm run pact",
-    "test:unit": "./node_modules/.bin/jest --detectOpenHandles --testPathIgnorePatterns=tests/pact",
+    "test:unit": "./node_modules/.bin/jest --detectOpenHandles --testPathIgnorePatterns=tests/pact tests/api",
```
Comment (sandymay-dd) [inline backend-api/package.json:15]:
```suggestion
    "test:unit": "./node_modules/.bin/jest --detectOpenHandles --testPathIgnorePatterns=tests/pact --testPathPatterns=src/,
```

File: backend-api/src/functions/tests/api/jwks.test.ts
Comment (jmooney-dd) [reply backend-api/src/functions/tests/api/jwks.test.ts:1]: Done

File: backend-api/template.yaml [hunk 2]
```diff
@@ -669,3 +667,204 @@ Resources:
     Properties:
       AliasName: !Sub alias/${AWS::StackName}-encryption-key
       TargetKeyId: !Ref KMSEncryptionKey
+
+  ### JWKS endpoint resources
+  JsonWebKeysBucket:
+    Type: AWS::S3::Bucket
+    Properties:
+      BucketName: !Sub ${AWS::StackName}-jwks-${Environment}
```
Comment (sandymay-dd) [inline backend-api/template.yaml:675]:
```suggestion
      BucketName: !Sub ${AWS::StackName}-jwks-keys
```

File: backend-api/template.yaml [hunk 3]
```diff
@@ -669,3 +667,204 @@ Resources:
...
+        BlockPublicPolicy: true
+        IgnorePublicAcls: true
+        RestrictPublicBuckets: true
+      BucketEncryption:
+        ServerSideEncryptionConfiguration:
+          - ServerSideEncryptionByDefault:
+              SSEAlgorithm: AES256
+      LoggingConfiguration:
+        DestinationBucketName: !Ref JsonWebKeysBucketAccessLogs
+        LogFilePrefix: !Sub ${AWS::StackName}-jwks-${Environment}
```
Comment (sandymay-dd) [inline backend-api/template.yaml:689]:
```suggestion
      LogFilePrefix: !Sub ${AWS::StackName}-jwks-keys
```

File: backend-api/template.yaml [hunk 4]
```diff
@@ -669,3 +667,204 @@ Resources:
...
+              - arn:aws:s3:::${bucketName}/*
+              - bucketName: !Ref JsonWebKeysBucket
+            Condition:
+              Bool:
+                "aws:SecureTransport": false
+
+  JsonWebKeysBucketAccessLogs:
+    Type: AWS::S3::Bucket
+    Properties:
+      BucketName: !Sub ${AWS::StackName}-jwks-${Environment}-access-logs
```
Comment (sandymay-dd) [inline backend-api/template.yaml:712]:
```suggestion
      BucketName: !Sub ${AWS::StackName}-jwks-keys-access-logs
```

**Review by sandymay-dd — DISMISSED** [review]

File: backend-api/template.yaml (hunk 4 above)
Comment (jmooney-dd) [reply backend-api/template.yaml]: Done

File: backend-api/template.yaml (hunk 3 above)
Comment (jmooney-dd) [reply backend-api/template.yaml]: Done

File: backend-api/template.yaml (hunk 2 above)
Comment (jmooney-dd) [reply backend-api/template.yaml]: Done

**Review by sandymay-dd — APPROVED** [review]

**Pull request merged**